### PR TITLE
[Sprint] Bug fixes for AST->ASR transition to compile FPM

### PR DESCRIPTION
**Pull request by czgdp1807**
Continues https://github.com/lfortran/lfortran/pull/1153

Commits to be added in `main`,

- [x] e767b428169b324e2f52d366aece8a79813bfe00 - https://github.com/lfortran/lfortran/pull/1273 
- [x] 81c32589527981e45aac6a66ac89348c130dbb18 - https://github.com/lfortran/lfortran/pull/1273
- [x] 12b1f899b4268df0f020165cb3184de2eb9414c8 - https://github.com/lfortran/lfortran/pull/1269
- [x] bf7aa415ec510106caa4b91c723baf0e1aadb88a - https://github.com/lfortran/lfortran/pull/1269
- [x] 056b4f2125a115bccec14b48d8b25c22451d8dbe - https://github.com/lfortran/lfortran/pull/1270
- [x] 08b8658d92aeff18053e485cb4cbfe293ec6d975 - https://github.com/lfortran/lfortran/pull/1270
- [x] 653267b9aeacad1adf72f12fcbb2af8655ee4965 - https://github.com/lfortran/lfortran/pull/1271
- [x] 19fe4305ce5321e7e7d218871eb5567a66fe3e26 - https://github.com/lfortran/lfortran/pull/1295
- [x] f03d356ba9a987fd59eb0ff721063c6385cde01a - https://github.com/lfortran/lfortran/pull/1272 
- [x] cc5dff2fe77d847d9df5c633d5cd3229ecdc33cc - https://github.com/lfortran/lfortran/pull/1272 
- [x] 4b6a482113e74da21bc5cacee154e0be22108f8d - https://github.com/lfortran/lfortran/pull/1272
- [x] edb65d4d1b650efea1629c6299714d26c7c43b57 - https://github.com/lfortran/lfortran/pull/1272
- [x] cecb51d0b9907c546a72e642f7301ee3b0d16018 - https://github.com/lfortran/lfortran/pull/1272
- [x] 8ccc71dd3184908d4ba946a230b4eba62d745983 - https://github.com/lfortran/lfortran/pull/1288 
- [x] 908b9f3cf3da9cddc4b514737896d9efc1c6c4d3 - https://github.com/lfortran/lfortran/pull/1288
- [x] 47eed37a39ff572eb6b69cd0862ef75adcecec14 - https://github.com/lfortran/lfortran/pull/1288
- [x] 6ea8b73020669248ad4f5a311e513876ddc900ae - https://github.com/lfortran/lfortran/pull/1276
- [x] feffbd9240120adfa063aa1adb532569c1acde9c - https://github.com/lfortran/lfortran/pull/1277 
- [x] 23e0921c868913559da4e655b094bb24750fe27e - https://github.com/lfortran/lfortran/pull/1296
- [x] 9fbd653edb1bd5979fae9553591cbd4343914fbd - https://github.com/lfortran/lfortran/pull/1289
- [x] 9116df333caa7135a0ef4ed56732418a5c7d50d8 - https://github.com/lfortran/lfortran/pull/1283
- [x] b6133a9c258e62f2e775da35621385f465a3430c - https://github.com/lfortran/lfortran/pull/1288 
- [x] cb0b767f7c724153aa4a2f9ac591e787a1f083e1 - https://github.com/lfortran/lfortran/pull/1296
- [x] 90817bd43f10b5f024e5af74dac49214f76e98ce - https://github.com/lfortran/lfortran/pull/1276
- [x] 579435d1f48081914bff0d361d25a04178013f41 - https://github.com/lfortran/lfortran/pull/1279 
- [x] 4bc78805431f0f95e570b4a9e429980022f5f42a - https://github.com/lfortran/lfortran/pull/1280 
- [x] 7ff56529e93b06cf38bcc3c5ed5672a9581f2f82 - https://github.com/lfortran/lfortran/pull/1280 
- [x] 7bb475f5b8f7cb5658b8bf1ce89a2b5e11801fa6 - https://github.com/lfortran/lfortran/pull/1277
- [x] a13c6703eb6cfeda7291ad5747feaa7979f9490f - https://github.com/lfortran/lfortran/pull/1288
- [x] 0e578dfa12bdedc8ed467b2f701d686659e2e58a - https://github.com/lfortran/lfortran/pull/1290 
- [x] 8728172dcb3471bbb437e5d3d26a99ed61b56507 - https://github.com/lfortran/lfortran/pull/1278
- [x] 5f583b8fb471fa296a993a5cad897b1913255232 - https://github.com/lfortran/lfortran/pull/1290
- [x] 4474612885bca5879c1614b7029a450207c60981 - https://github.com/lfortran/lfortran/pull/1281 
- [x] 9b4c5c6e0daa762275df38783d9ab561989e2466 - https://github.com/lfortran/lfortran/pull/1279
- [x] 6df1d0c2665b379d222030ba65d2d978dff89c0b - https://github.com/lfortran/lfortran/pull/1281 
- [x] 282c93cf249aceb9f2d12c4266c500d04c67e467 - https://github.com/lfortran/lfortran/pull/1282 

@Thirumalai-Shaktivel @Smit-create Please continue your sprint here. Also please do not rebase on top of main branch. I need the commits to cherry-pick in future, so **please do not change the git history of the PR**. Just add new commits and push.

Progress:

```bash
(lf) 16:13:11:~/lfortran_project/fpm % FPM_FC="lfortran" fpm build --flag="--cpp"
filesystem_utilities.c                 done.
fpm_strings.f90                        done.
fpm_backend_console.f90                done.
iscygpty.c                             done.
isatty.c                               done.
constants.f90                          done.
version.f90                            done.
M_CLI2.F90                             done.
error.f90                              done.
error.f90                              done.
datetime.f90                           done.
verify.f90                             done.
fpm_environment.f90                    done.
fpm_os.F90                             done.
versioning.f90                         done.
convert.f90                            done.
fpm_filesystem.F90                     done.
utils.f90                              done.
fpm_command_line.f90                   done.
installer.f90                          done.
git.f90                                done.
value.f90                              done.
sort.f90                               done.
keyval.f90                             done.
base.f90                               done.
vector.f90                             done.
structure.f90                          done.
table.f90                              done.
array.f90                              done.
type.f90                               done.
ser.f90                                done.
tokenizer.f90                          done.
merge.f90                              done.
keyval.f90                             done.
character.f90                          done.
table.f90                              done.
array.f90                              done.
de.f90                                 done.
build.f90                              done.
tomlf.f90                              done.
toml.f90                               done.
profiles.f90                           done.
install.f90                            done.
preprocess.f90                         done.
build.f90                              done.
dependency.f90                         done.
library.f90                            done.
executable.f90                         done.
test.f90                               done.
example.f90                            done.
package.f90                            done.
manifest.f90                           done.
new.f90                                done.
fpm_compiler.f90                       done.
dependency.f90                         done.
fpm_model.f90                          done.
update.f90                             done.
fpm_source_parsing.f90                 done.
fpm_targets.f90                        done.
fpm_backend_output.f90                 done.
fpm_sources.f90                        done.
fpm_backend.F90                        done.
fpm.f90                                done.
install.f90                            done.
libfpm.a                               done.
main.f90                               failed.
[ 98%] Compiling...
ASR verify pass error: ASR verify: Function get_command_line_settings doesn't depend on any but is found in its dependency list.
Internal Compiler Error: Unhandled exception
Traceback (most recent call last):
  File "/Users/czgdp1807/lfortran_project/lfortran/src/bin/lfortran.cpp", line 1907
    return compile_to_object_file(arg_file, outfile, false,
  File "/Users/czgdp1807/lfortran_project/lfortran/src/bin/lfortran.cpp", line 865
    res = fe.get_llvm3(*asr, lpm, diagnostics, infile);
  File "/Users/czgdp1807/lfortran_project/lfortran/src/lfortran/fortran_evaluator.cpp", line 342
    compiler_options, run_fn, infile);
  File "/Users/czgdp1807/lfortran_project/lfortran/src/libasr/codegen/asr_to_llvm.cpp", line 6794
    pass_manager.apply_passes(al, &asr, pass_options, diagnostics);
  File "/Users/czgdp1807/lfortran_project/lfortran/src/libasr/pass/pass_manager.h", line 209
    _apply_passes(al, asr, _passes, pass_options, diagnostics);
  File "/Users/czgdp1807/lfortran_project/lfortran/src/libasr/pass/pass_manager.h", line 106
    throw LCompilersException("Verify failed");
LCompilersException: Verify failed
<ERROR> Compilation failed for object " app_main.f90.o "
<ERROR>stopping due to failed compilation
STOP 1
```

### CODE REVIEW

**Comment by czgdp1807** [comment]
@Smit-create Please continue the sprint. I have progressed till 61%.

File: src/lfortran/semantics/ast_symboltable_visitor.cpp [hunk 1]
```diff
@@ -499,13 +499,15 @@ class SymbolTableVisitor : public CommonVisitor<SymbolTableVisitor> {
             }
         }
         if (parent_scope->get_symbol(sym_name) != nullptr) {
-            ASR::symbol_t *f1 = parent_scope->get_symbol(sym_name);
+            ASR::symbol_t *f1 = ASRUtils::symbol_get_past_external(
+                parent_scope->get_symbol(sym_name));
```
Comment (Thirumalai-Shaktivel) [inline src/lfortran/semantics/ast_symboltable_visitor.cpp:503]: @czgdp1807 This PR: https://github.com/lfortran/lfortran/pull/1204/files was created to fix this issue.
Shall I apply the following changes there and mark them for review?

Comment (czgdp1807) [reply src/lfortran/semantics/ast_symboltable_visitor.cpp:503]: Please don’t make any change to this PR or its git history. However feel free to go ahead and modify your PRs by applying changes from here. Thanks.

File: examples/expr2.f90 [hunk 2]
```diff
@@ -1,9 +1,12 @@
...
+module debug_1
+    use debug_2
+    implicit none
+    type t_1
+        type(t_2) :: type_1
+    end type t_1
+contains
+    subroutine s_1(self)
+        type(t_1) :: self
+        self%type_1 = f_2()
```
[the remark this answers is not in the document]
Comment (Thirumalai-Shaktivel) [reply examples/expr2.f90:10]: Is `FunctionCall` modified by some replacer after `ast_common_visitor`?

[the remark this answers is not in the document]
Comment (czgdp1807) [reply examples/expr2.f90:10]: Thanks for the update. I will do the sprint tomorrow and will fix the issue. Please push all your commits and I will carry on tomorrow.

[the remark this answers is not in the document]
Comment (Thirumalai-Shaktivel) [reply examples/expr2.f90:10]: I worked only on this issue, and I pushed it here.

[the remark this answers is not in the document]
Comment (Thirumalai-Shaktivel) [reply examples/expr2.f90:10]: I think this also has the similar issue
https://github.com/lfortran/lfortran/blob/6ee8f6226e62c09684110bf2fef7c3f74e3239eb/integration_tests/modules_24.f90#L36-L40

**Comment by czgdp1807** [comment]
@Thirumalai-Shaktivel Please continue the sprint tomorrow. I have made progress till 79%.

**Comment by czgdp1807** [comment]
@Thirumalai-Shaktivel Please feel free continue the sprint. I will be out of office tomorrow. I have progressed till 88%.

**Comment by certik** [comment]
Nice, almost done!!

**Comment by Thirumalai-Shaktivel** [comment]
Next example on which I was working on:
```fortran
module module_1
    implicit none
contains
subroutine sub_1(a, b, c, d)
    integer, intent(in) :: a
    integer, intent(in), optional :: b
    integer, intent(in), optional :: c
    integer, intent(in) :: d
end subroutine sub_1

end module module_1

module module_2
    use module_1, only: sub_1
contains

    subroutine sub_2()
        call sub_1(1, b=1, c=1, d=2)
        ! call sub_1(1, d=2)
    end subroutine sub_2
end module module_2
```

**Comment by czgdp1807** [comment]
So compilation till ASR is complete. As you can see in the description of this PR, the error is in LLVM backend now. So, I would mark this sprint as complete. We will know cherry-pick commits and apply to the PR. Later today, I will update the description of this PR with list of commits (and assign to @Smit-create, @Thirumalai-Shaktivel and myself) to be merged into main.

**Comment by certik** [comment]
Yes, let's merge everything into main and setup a CI to compile all of fpm into ASR, to ensure this keeps working.

File: src/libasr/asr_utils.h [hunk 3]
```diff
@@ -2118,18 +2128,105 @@ static inline ASR::ttype_t* get_type_parameter(ASR::ttype_t* t) {
...
+        SymbolTable* import_struct_t_scope = scope;
+        while( import_struct_t_scope->asr_owner == nullptr ||
+               !(ASR::is_a<ASR::Module_t>(*ASR::down_cast<ASR::symbol_t>(
+                import_struct_t_scope->asr_owner))) ) {
+            import_struct_t_scope = import_struct_t_scope->parent;
+            if( import_struct_t_scope->asr_owner != nullptr &&
+                !ASR::is_a<ASR::symbol_t>(*import_struct_t_scope->asr_owner) ) {
+                break;
+            }
+        }
```
Comment (czgdp1807) [inline src/libasr/asr_utils.h:2154]: Added in https://github.com/lfortran/lfortran/pull/1269

**Comment by certik** [comment]
@Thirumalai-Shaktivel, @Smit-create  can you please create PRs for your commits here? Let's get everything into master and set it up at a CI to test full fpm to ASR.

**Comment by Smit-create** [comment]
Sure, I'll complete it by tomorrow. Thanks!

**Comment by certik** [comment]
Looks like we have 3 PRs left above, let's finish it as soon as we can, so that we can enable the CI and after that we can work on refactoring ASR (as needed) and ASR->LLVM.

**Comment by certik** [comment]
I tested fpm version 9cb38b2a9f3c31d6617a48e757cd56556b0d05b5, LFortran in Release mode on Apple M1 Max:
```console
$ time FPM_FC="lfortran" fpm build --flag="--cpp"
filesystem_utilities.c                 done.
fpm_strings.f90                        done.
fpm_backend_console.f90                done.
iscygpty.c                             done.
isatty.c                               done.
constants.f90                          done.
version.f90                            done.
M_CLI2.f90                             done.
error.f90                              done.
error.f90                              done.
datetime.f90                           done.
verify.f90                             done.
fpm_environment.f90                    done.
fpm_os.F90                             done.
versioning.f90                         done.
convert.f90                            done.
fpm_filesystem.F90                     done.
utils.f90                              done.
fpm_command_line.f90                   done.
installer.f90                          done.
git.f90                                done.
value.f90                              done.
keyval.f90                             done.
base.f90                               done.
vector.f90                             done.
structure.f90                          done.
table.f90                              done.
array.f90                              done.
type.f90                               done.
ser.f90                                done.
tokenizer.f90                          done.
merge.f90                              done.
keyval.f90                             done.
character.f90                          done.
table.f90                              done.
array.f90                              done.
de.f90                                 done.
build.f90                              done.
tomlf.f90                              done.
toml.f90                               done.
profiles.f90                           done.
install.f90                            done.
preprocess.f90                         done.
build.f90                              done.
dependency.f90                         done.
library.f90                            done.
executable.f90                         done.
test.f90                               done.
example.f90                            done.
package.f90                            done.
manifest.f90                           done.
new.f90                                done.
fpm_compiler.f90                       done.
dependency.f90                         done.
fpm_model.f90                          done.
update.f90                             done.
fpm_source_parsing.f90                 done.
fpm_targets.f90                        done.
fpm_backend_output.f90                 done.
fpm_sources.f90                        done.
fpm_backend.F90                        done.
fpm.f90                                done.
install.f90                            done.
libfpm.a                               done.
main.f90                               compiling...
[ 96%] Compiling...sh: line 1: 11129 Segmentation fault: 11  lfortran -c app/main.f90 --cpp -J build/lfortran_BC003944AE0DF3D6 -Ibuild/lfortran_BC003944AE0DF3D6 -o build/lfortran_BC003944AE0DF3D6/fpm/app_main.f90.o > build/lfortran_BC003944main.f90                               failed.
[ 98%] Compiling...
<ERROR> Compilation failed for object " app_main.f90.o "
<ERROR>stopping due to failed compilation
STOP 1
FPM_FC="lfortran" fpm build --flag="--cpp"  1.02s user 0.27s system 92% cpu 1.402 total
```

And GFortran 11.3.0
```console
$ time fpm build
filesystem_utilities.c                 done.
fpm_strings.f90                        done.
fpm_backend_console.f90                done.
iscygpty.c                             done.
isatty.c                               done.
constants.f90                          done.
version.f90                            done.
M_CLI2.f90                             done.
error.f90                              done.
error.f90                              done.
datetime.f90                           done.
verify.f90                             done.
fpm_environment.f90                    done.
fpm_os.F90                             done.
versioning.f90                         done.
convert.f90                            done.
fpm_filesystem.F90                     done.
utils.f90                              done.
fpm_command_line.f90                   done.
installer.f90                          done.
git.f90                                done.
value.f90                              done.
keyval.f90                             done.
base.f90                               done.
vector.f90                             done.
structure.f90                          done.
table.f90                              done.
array.f90                              done.
type.f90                               done.
ser.f90                                done.
tokenizer.f90                          done.
merge.f90                              done.
keyval.f90                             done.
character.f90                          done.
table.f90                              done.
array.f90                              done.
de.f90                                 done.
build.f90                              done.
tomlf.f90                              done.
toml.f90                               done.
profiles.f90                           done.
install.f90                            done.
preprocess.f90                         done.
build.f90                              done.
dependency.f90                         done.
library.f90                            done.
executable.f90                         done.
test.f90                               done.
example.f90                            done.
package.f90                            done.
manifest.f90                           done.
new.f90                                done.
fpm_compiler.f90                       done.
dependency.f90                         done.
fpm_model.f90                          done.
update.f90                             done.
fpm_source_parsing.f90                 done.
fpm_targets.f90                        done.
fpm_backend_output.f90                 done.
fpm_sources.f90                        done.
fpm_backend.F90                        done.
fpm.f90                                done.
install.f90                            done.
libfpm.a                               done.
main.f90                               done.
fpm                                    done.
[100%] Project compiled successfully.
fpm build  6.36s user 0.81s system 95% cpu 7.530 total
```
So LFortran is 5.3x faster to compile, so far. We'll see how much extra the ASR->LLVM adds. It looks promising.

**Comment by certik** [comment]
TODO:
* [ ] after the last PR above is in, we should setup a CI using the script above
* [ ] we will shift to ASR->LLVM

**Comment by certik** [comment]
With the following patch:
```diff
--- a/src/bin/lfortran.cpp
+++ b/src/bin/lfortran.cpp
@@ -849,16 +849,16 @@ int compile_to_object_file(const std::string &infile,
     }
 
     // ASR -> LLVM
-    LCompilers::LLVMEvaluator e(compiler_options.target);
 
     if (!compiler_options.generate_object_code
             && !LCompilers::ASRUtils::main_program_present(*asr)) {
         // Create an empty object file (things will be actually
         // compiled and linked when the main program is present):
-        e.create_empty_object_file(outfile);
+        //e.create_empty_object_file(outfile);
         return 0;
     }
 
+    LCompilers::LLVMEvaluator e(compiler_options.target);
     std::unique_ptr<LCompilers::LLVMModule> m;
     diagnostics.diagnostics.clear();
     LCompilers::Result<std::unique_ptr<LCompilers::LLVMModule>>
```
Which omits generating the empty object file, I now get:
```console
$ time ./build.sh
...
+ lfortran -c ././src/fpm/cmd/install.f90 --cpp -J build/lfortran_BC003944AE0DF3D6 -Ibuild/lfortran_BC003944AE0DF3D6 -o build/lfortran_BC003944AE0DF3D6/fpm/src_fpm_cmd_install.f90.o
+ lfortran --show-asr app/main.f90 --cpp -J build/lfortran_BC003944AE0DF3D6 -Ibuild/lfortran_BC003944AE0DF3D6
./build.sh  0.85s user 0.16s system 94% cpu 1.060 total
```
Which is about 6.5x faster.

**Comment by czgdp1807** [comment]
All clear here. Closing this.

**Pull request closed without merging**